File: prolog/2024/2024-004/xmas_counter.pl
:- module(xmas_counter, [main/0]).

% Read the grid from the input file
read_grid(FilePath, Grid) :-
    setup_call_cleanup(
        open(FilePath, read, Stream),
        read_lines(Stream, Grid),
        close(Stream)
    ).

read_lines(Stream, Lines) :-
    read_string(Stream, _, Content),
    split_string(Content, "\n", "\s\t", Lines).

% Check if a word exists in a given direction
check_word(Grid, X, Y, Dx, Dy, Word) :-
    string_length(Word, Len),
    length(Grid, Rows),
    nth0(0, Grid, FirstRow),
    string_length(FirstRow, Cols),
    check_word(Grid, X, Y, Dx, Dy, Word, Len, Rows, Cols, 0).

check_word(_, _, _, _, _, _, Len, _, _, Len).
check_word(Grid, X, Y, Dx, Dy, Word, Len, Rows, Cols, I) :-
    Nx is X + I * Dx,
    Ny is Y + I * Dy,
    Nx >= 0, Ny >= 0, Nx < Rows, Ny < Cols,
    nth0(Nx, Grid, Row),
    sub_atom(Row, Ny, 1, _, GridChar),
    sub_atom(Word, I, 1, _, WordChar),
    GridChar == WordChar,
    NextI is I + 1,
    check_word(Grid, X, Y, Dx, Dy, Word, Len, Rows, Cols, NextI).

% Count occurrences of the word "XMAS" in all directions
count_xmas(Grid, Count) :-
    TargetWord = "XMAS",
    Directions = [(0, 1), (1, 0), (1, 1), (1, -1), (0, -1), (-1, 0), (-1, -1), (-1, 1)],
    findall(1, (
        nth0(R, Grid, _),
        nth0(C, Grid, _),
        member((Dx, Dy), Directions),
        check_word(Grid, R, C, Dx, Dy, TargetWord)
    ), Matches),
    length(Matches, Count).

% Count all X-MAS patterns in the grid
count_all_xmas_patterns(Grid, Count) :-
    findall(1, (
        nth0(R, Grid, _),
        nth0(C, Grid, _),
        valid_xmas_pattern(Grid, R, C)
    ), Matches),
    length(Matches, Count).

valid_xmas_pattern(Grid, R, C) :-
    R > 0, C > 0,
    length(Grid, Rows),
    nth0(0, Grid, FirstRow),
    string_length(FirstRow, Cols),
    R < Rows - 1, C < Cols - 1,  % Ensure bounds for neighbors
    R1 is R - 1,  % Compute row above
    R2 is R + 1,  % Compute row below
    C1 is C - 1,  % Compute column left
    C2 is C + 1,  % Compute column right
    (
        % M.S \n A \n M.S
        (nth0(R1, Grid, Row1), sub_atom(Row1, C1, 1, _, 'M'), sub_atom(Row1, C2, 1, _, 'S'),
         nth0(R, Grid, Row2), sub_atom(Row2, C, 1, _, 'A'),
         nth0(R2, Grid, Row3), sub_atom(Row3, C1, 1, _, 'M'), sub_atom(Row3, C2, 1, _, 'S'));
        % S.M \n A \n S.M
        (nth0(R1, Grid, Row1), sub_atom(Row1, C1, 1, _, 'S'), sub_atom(Row1, C2, 1, _, 'M'),
         nth0(R, Grid, Row2), sub_atom(Row2, C, 1, _, 'A'),
         nth0(R2, Grid, Row3), sub_atom(Row3, C1, 1, _, 'S'), sub_atom(Row3, C2, 1, _, 'M'));
        % M.M \n A \n S.S
        (nth0(R1, Grid, Row1), sub_atom(Row1, C1, 1, _, 'M'), sub_atom(Row1, C2, 1, _, 'M'),
         nth0(R, Grid, Row2), sub_atom(Row2, C, 1, _, 'A'),
         nth0(R2, Grid, Row3), sub_atom(Row3, C1, 1, _, 'S'), sub_atom(Row3, C2, 1, _, 'S'));
        % S.S \n A \n M.M
        (nth0(R1, Grid, Row1), sub_atom(Row1, C1, 1, _, 'S'), sub_atom(Row1, C2, 1, _, 'S'),
         nth0(R, Grid, Row2), sub_atom(Row2, C, 1, _, 'A'),
         nth0(R2, Grid, Row3), sub_atom(Row3, C1, 1, _, 'M'), sub_atom(Row3, C2, 1, _, 'M'))
    ).



% Main function
main :-
    FilePath = "input.txt",
    read_grid(FilePath, Grid),
    (Grid = [] ->
        writeln("Error: Grid is empty or invalid.");
        (
            count_xmas(Grid, XmasCount),
            format("Count of XMAS: ~w~n", [XmasCount]),
            count_all_xmas_patterns(Grid, XmasPatterns),
            format("Total X-MAS patterns: ~w~n", [XmasPatterns])
        )
    ).

File: other-programming-languages/Erlang/2024/2024-004/xmas_counter.erl
-module(xmas_counter).
-export([main/0, read_grid/1, count_xmas/1, count_all_xmas_patterns/1, check_word/6]).

%% Read the grid from the input file
read_grid(FilePath) ->
    {ok, Binary} = file:read_file(FilePath),
    Lines = binary:split(Binary, <<"\n">>, [global, trim]),
    [binary_to_list(Line) || Line <- Lines].

%% Check if a word exists in a given direction
check_word(Grid, X, Y, Dx, Dy, Word) ->
    Len = length(Word),
    Rows = length(Grid),
    Cols = length(hd(Grid)),
    check_word(Grid, X, Y, Dx, Dy, Word, Len, Rows, Cols, 0).

check_word(_Grid, _X, _Y, _Dx, _Dy, _Word, Len, _Rows, _Cols, Len) ->
    true;
check_word(Grid, X, Y, Dx, Dy, Word, Len, Rows, Cols, I) ->
    Nx = X + I * Dx,
    Ny = Y + I * Dy,
    if
        Nx < 0; Ny < 0; Nx >= Rows; Ny >= Cols ->
            false;
        true ->
            GridRow = lists:nth(Nx + 1, Grid),
            GridChar = lists:nth(Ny + 1, GridRow),
            WordChar = lists:nth(I + 1, Word),
            case GridChar =/= WordChar of
                true -> false;
                false -> check_word(Grid, X, Y, Dx, Dy, Word, Len, Rows, Cols, I + 1)
            end
    end.

%% Count occurrences of the word "XMAS" in all directions
count_xmas(Grid) ->
    TargetWord = "XMAS",
    Directions = [{0, 1}, {1, 0}, {1, 1}, {1, -1}, {0, -1}, {-1, 0}, {-1, -1}, {-1, 1}],
    Rows = length(Grid),
    Cols = length(hd(Grid)),
    lists:sum([1 || R <- lists:seq(0, Rows - 1),
                   C <- lists:seq(0, Cols - 1),
                   {Dx, Dy} <- Directions,
                   check_word(Grid, R, C, Dx, Dy, TargetWord)]).

%% Count all X-MAS patterns in the grid
count_all_xmas_patterns(Grid) ->
    Rows = length(Grid),
    Cols = length(hd(Grid)),
    lists:sum([
        1 || R <- lists:seq(1, Rows - 2),
             C <- lists:seq(1, Cols - 2),
             lists:nth(C + 1, lists:nth(R + 1, Grid)) =:= $A,
             (
                 lists:nth(C, lists:nth(R, Grid)) =:= $M andalso
                 lists:nth(C + 2, lists:nth(R, Grid)) =:= $S andalso
                 lists:nth(C, lists:nth(R + 2, Grid)) =:= $M andalso
                 lists:nth(C + 2, lists:nth(R + 2, Grid)) =:= $S
             ) orelse
             (
                 lists:nth(C, lists:nth(R, Grid)) =:= $S andalso
                 lists:nth(C + 2, lists:nth(R, Grid)) =:= $M andalso
                 lists:nth(C, lists:nth(R + 2, Grid)) =:= $S andalso
                 lists:nth(C + 2, lists:nth(R + 2, Grid)) =:= $M
             ) orelse
             (
                 lists:nth(C, lists:nth(R, Grid)) =:= $M andalso
                 lists:nth(C + 2, lists:nth(R, Grid)) =:= $M andalso
                 lists:nth(C, lists:nth(R + 2, Grid)) =:= $S andalso
                 lists:nth(C + 2, lists:nth(R + 2, Grid)) =:= $S
             ) orelse
             (
                 lists:nth(C, lists:nth(R, Grid)) =:= $S andalso
                 lists:nth(C + 2, lists:nth(R, Grid)) =:= $S andalso
                 lists:nth(C, lists:nth(R + 2, Grid)) =:= $M andalso
                 lists:nth(C + 2, lists:nth(R + 2, Grid)) =:= $M
             )
    ]).



%% Main function
main() ->
    FilePath = "input.txt",
    Grid = read_grid(FilePath),
    case Grid of
        [] ->
            io:format("Error: Grid is empty or invalid.~n");
        _ ->
            XmasCount = count_xmas(Grid),
            io:format("Count of XMAS: ~p~n", [XmasCount]),
            XmasPatterns = count_all_xmas_patterns(Grid),
            io:format("Total X-MAS patterns: ~p~n", [XmasPatterns])
    end.
